R/featEng-tails.R

```diff
@@ -0,0 +1,166 @@
+#' is assumed to be an index column counting from 1 to \code{nrow(voxel_df)},
+#' and the second column is a column labeled "PID," which contains patient
+#' identification. The third, fourth, and fifth columns are 'x', 'y', and 'z'
+#' columns. The sixth column contains the value of the voxel that corresponds
+#' with the given x-y-z coordinate.
+#' @param data_df A \code{tibble} that is meant to be outputted from \code{dataDrivenClusters()}.
+#' This tibble should be in the following format: the first three columns are
+#' assumed to be 'x', 'y', and 'z' coordinates. The fourth and fifth columns are
+#' UMAP coordinates, and the fifth column contains cluster labels. If this is
+#' equal to \code{null}, \code{computeTailMeans()} will assume that all data comes
+#' from the same cluster.
+#' @param alpha A \code{double} that indicates what quantiles should be extracted
+#' from the patient data. The lower the number, the tighter the tail will be.
+#'
+#' @export
+
+computeTailMeans <- function(voxel_df, data_df = NULL, alpha = 0.05) {
+
+  # assuming long format voxel_df in which column 1 is index, column 2 is PID,
+  # column 3-5 is x,y,z, and column 6 is value. Assuming data_df to be output
+  # from dataDrivenClusters(), in which columns 1-3 are x,y,z, column 4-5 are
+  # UMAP coords, and column 6 is cluster labels.
+
+  if (alpha != 0 & alpha != 1) {
+
+    if (!is.null(data_df)) {
+
+      # cbind voxel_df and data_df
+      combine_df <- cbind(voxel_df, data_df)
+
+      # extracting the columns in following order: PID, x, y, z, Value, Cluster
+      combind_df <- combine_df[, c(2, 3, 4, 5, 6, 12)]
+
+      # adding additional xyz identifier column
+      combine_df <- combine_df %>% mutate(xyz = paste0(combine_df[, 1],", ",combine_df[, 2],", ",combine_df[, 3]))
+
+      #removing x, y, z columns since those are redundant after the xyz column.
+      combine_df <- combine_df[, c(1, 5, 6, 7)]
+
+      num_clusts <- length(unique(combine_df[, 2]))
+
+      # initializing empty vectors
+      clust_vec <- c()
+      PID_vec <- c()
+      lower_mean_vec <- c()
+      upper_mean_vec <- c()
+      lower_quantile_vec <- c()
+      upper_quantile_vec <- c()
+
+      # cycling through each cluster label
+      for (i in 1:num_clusts) {
+
+        # isolating individual clusters
+        filtered_df <- combine_df %>% filter(cluster == i)
+
+          # for each cluster, isolate individual PID
+          for(ID in unique(combine_df[, 1])) {
+
+          # filter the dataframe so that we are only considering individual PIDs
+          uniqueid_df <- filtered_df %>% filter(pid == ID)
+
+          # find quantile cutoffs
+          quantile_cutoff_lower <- quantile(uniqueid_df[, 3], alpha)
+          quantile_cutoff_upper <- quantile(uniqueid_df[, 3], 1-alpha)
+
+          # find values that fall beneath and above quantile cutoffs
+          values_lower <- uniqueid_df[, 3][which(uniqueid_df[, 3] < quantile_cutoff_lower)]
```

Comment:
You should use `dplyr::group_by(pid)` combined with `dplyr::top_frac(alpha)` or `dplyr::top_frac(-alpha)` instead.

`group_by(pid)` groups by patients such that the top fraction is with respect to each patient as opposed to the entire dataset

See my code or https://dplyr.tidyverse.org/reference/top_n.html for references, or DM me if this is unclear!